core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -1770,12 +1763,17 @@ private[zio] trait ZIOFunctions extends Serializable {
   final def foreachParN[R, E, A, B](
     n: Long
   )(as: Iterable[A])(fn: A => ZIO[R, E, B]): ZIO[R, E, List[B]] =
-    (for {
-      semaphore <- Semaphore.make(n)
-      bs <- foreachPar[R, E, A, B](as) { a =>
-             semaphore.withPermit(fn(a))
-           }
-    } yield bs).refailWithTrace
+    for {
+      q     <- Queue.bounded[(Promise[E, B], A)](n.toInt)
+      pairs <- ZIO.foreach(as)(a => Promise.make[E, B].map(p => (p, a)))
+      _     <- ZIO.foreach(pairs)(pair => q.offer(pair)).fork
+      _ <- ZIO.collectAll(
+            List.fill(n.toInt)(
+              q.take.flatMap { case (p, a) => fn(a).onError(_ => q.shutdown).flatMap(p.succeed) }.forever.fork
+            )
+          )
+      res <- ZIO.collectAll(pairs.map(_._1.await)).ensuring(q.shutdown)
```

Comment:
This could be rewritten using ZIO.foreach

Reply:
Yes.

Reply:
Got it. It seems using a collection size of 1,000,000 elements in the tests was causing some flakiness (at different times we were failing on 2.13 / JDK 8, 2.13 / JDK 11, and 2.12 / JDK 8 in what appeared to be unrelated tests) so I reduced it to 100,000 and everything is working now. What do you think?